cassiopeia/simulator/BirthDeathFitnessSimulator.py

```diff
@@ -0,0 +1,336 @@
+        tree.add_node(0)
+        tree.nodes[0]["birth_scale"] = birth_scale_param
+        tree.nodes[0]["total_time"] = 0
+        current_lineages = []
+        starting_lineage = {
+            "id": 0,
+            "birth_scale": birth_scale_param,
+            "total_time": 0,
+        }
+
+        # Sample the waiting time until the first division
+        unique_id = 1
+        unique_id = sample_event(unique_id, starting_lineage)
+
+        # Perform the process until there are no active extant lineages left
+        while len(current_lineages) > 0:
+            # If number of extant lineages is the stopping criterion, at the
+            # first instance of having n extant tips, stop the experiment
+            # and set the total lineage time for each lineage to be equal to
+            # the minimum, to produce ultrametric trees. Also, the birth_scale
+            # parameter of each leaf is rolled back to equal its parent's.
+            if num_extant:
+                if len(current_lineages) == num_extant:
+                    min_total_time = min(
+                        [i["total_time"] for i in current_lineages]
+                    )
+                    for remaining_lineage in current_lineages:
+                        parent = list(
+                            tree.predecessors(remaining_lineage["id"])
+                        )[0]
+                        tree.edges[parent, remaining_lineage["id"]][
+                            "weight"
+                        ] += (min_total_time - remaining_lineage["total_time"])
+                        tree.nodes[remaining_lineage["id"]][
+                            "birth_scale"
+                        ] = tree.nodes[parent]["birth_scale"]
+                    break
+            # If extant tips are the stopping criteria, pop the minimum age
+            # lineage at each step
+            if num_extant:
+                min_age_ind = np.argmin(
+                    [i["total_time"] for i in current_lineages]
+                )
+                lineage = current_lineages.pop(min_age_ind)
+            else:
+                lineage = current_lineages.pop(0)
+            for _ in range(2):
+                unique_id = sample_event(unique_id, lineage)
+
+        # Prune dead lineages and collapse resulting unifurcations
+        if death_waiting_dist and len(tree.nodes) > 1:
+            if experiment_time:
+                for i in list(tree.nodes):
+                    if (
+                        tree.out_degree(i) == 0
+                        and tree.nodes[i]["total_time"] < experiment_time
+                    ):
+                        self.remove_and_prune_lineage(i, tree)
+            if num_extant:
+                surviving_ids = [i["id"] for i in current_lineages]
+                for i in list(tree.nodes):
+                    if tree.out_degree(i) == 0 and i not in surviving_ids:
+                        self.remove_and_prune_lineage(i, tree)
```

Comment:
For testing purposes it might be really cool to have an option to *not* prune the dead lineages, so that we can see where they are and that they look correct. I think an argument `prune_dead_lineages` in `__init__` would be ideal? However, this might not be easy to get working because I think `CassiopeiaTree` assumes that all leaves in the tree are extant...

One thing we *can* easily do though is to allow not collapsing unifurcations. That way, at least we get to see where in the tree the lineages that got extinct branched off. Again, this is only for testing / visualization purposes.